http-gateway/pkg/server/topic_existence_checker.go

```diff
@@ -33,29 +45,57 @@ func NewAlwaysTrueTopicExistenceChecker() TopicExistenceChecker {
 // NewRiffTopicExistenceChecker configures a TopicExistenceChecker using the
 // provided Clientset.
 func NewRiffTopicExistenceChecker(clientSet *versioned.Clientset) TopicExistenceChecker {
-	return &riffTopicExistenceChecker{client: clientSet}
-}
+	riffInformerFactory := informers.NewSharedInformerFactory(clientSet, time.Second*30)
+	topicInformer := riffInformerFactory.Projectriff().V1alpha1().Topics()
+	knownTopics := make(map[string]ignoredValue)
+	mutex := &sync.Mutex{}
+
+	topicInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			mutex.Lock()
+			defer mutex.Unlock()
 
-func (tec *alwaysTrueTopicExistenceChecker) TopicExists(namespace string, topicName string) (bool, error) {
-	return true, nil
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err != nil {
+				log.Printf("AddFunc had an error for key '%s': %+v", err)
+			}
+
+			knownTopics[key] = ignoredValue{}
```

Comment:
According to its docs, `MetaNamespaceKeyFunc` returns a string of the form `namespace/name` or of the form `name` if the namespace is empty (whatever that means - default? unknown?). This opens up the possibility of a bug in this implementation. If the namespace can be empty in situations when it is not the default, then the same topic name in two distinct namespaces could end up being represented by the same key and the map would lose track of the existence of both topics.

Reply:
For the moment I'm inclined to kick the missing-namespace case down the road, as my reckoning is that topics will always have namespaces.

Looking at the `cache` package where `MetaNamespaceKeyFunc` lives shows that it's the default implementation for `KeyFunc` function type. That might actually be the ideal extension point for us to implement checking across all namespaces instead of just the default namespace. It would also let us sidestep the missing-namespace issue you've raised here.

The `DeletionHandlingMetaNamespaceKeyFunc` (which implements `KeyFunc`) has a hardcoded dependency on `MetaNamespaceKeyFunc`, so we would need to replace that also.

I'll add a TODO.